### ISSUE TITLE
Return type compatibility improperly enforced in explicit Protocol implementations

### ISSUE THREAD

**Issue by mirhahn** [microsoft/pyright#5117]
**Description**
Given a custom protocol with a method with return type annotation and a class that inherits from the protocol (making it an explicit implementation), PyRight will issue neither an error nor a warning if the return type of the implementation's method is broader or otherwise incompatible with the protocol's return type. However, it will do so normally if the class is an not an explicit implementation.

**Reproduction**
Copy the following code into a file (e.g., `my_code.py`) and execute `pyright my_code.py`:
```python
from typing import Protocol, Literal


class MyProtocol(Protocol):
    def foo(self) -> Literal['a']:
        ...

    def bar(self) -> tuple[float]:
        ...


class MyClass(MyProtocol):
    # While this has a `Literal` annotation, it returns a literal not permitted by the protocol.
    def foo(self) -> Literal['b']:
        return 'b'

    # This returns a completely different type from the one given in the protocol.
    def bar(self) -> str:
        return 'a'


# This is a test function to see if PyRight will complain if a `MyClass` is passed to `func`.
def func(obj: MyProtocol):
    if obj.foo() == 'a':
        print('Something expected happened.')
    else:
        print('Something unexpected happened.')


me = MyClass()
func(me)
```

This should yield the following output:

```
$ pyright my_code.py
0 errors, 0 warnings, 0 informations
```

By removing the inheritance clause `(MyProtocol)` from `MyClass`, PyRight can be made to detect the error:

```
$ pyright my_implicit_code.py
[redacted]/my_implicit_code.py
  [redacted]/my_implicit_code.py:28:6 - error: Argument of type "MyClass" cannot be assigned to parameter "obj" of type "MyProtocol" in function "func"
    "MyClass" is incompatible with protocol "MyProtocol"
      "foo" is an incompatible type
        Type "() -> Literal['b']" cannot be assigned to type "() -> Literal['a']"
          Function return type "Literal['b']" is incompatible with type "Literal['a']"
            "Literal['b']" cannot be assigned to type "Literal['a']"
      "bar" is an incompatible type
        Type "() -> str" cannot be assigned to type "() -> tuple[float]"
          Function return type "str" is incompatible with type "tuple[float]"
    ... (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
``` 

**Expected behavior**
The error `"MyClass" is incompatible with protocol "MyProtocol"` and its sub-errors `"foo" is an incompatible type` and `"bar" is an incompatible type` should appear in the case of the explicit protocol implementation as well. They could appear as errors on the individual method definitions or in the first line of `MyClass` in a manner similar to how missing method errors appear.

**Version Information**
I am running PyRight version `1.1.308` using the command line interface. My Python version is `3.11.2`.

**Additional Context**
It is conceivable that PyRight reads an explicit implementation as the programmer asserting compliance with the protocol and overriding the checks. However, this appears implausible because PyRight does issue errors if an explicit implementation is missing methods from the protocol.

**Comment by erictraut** [comment]
In the case of explicit protocol implementations, pyright treats methods in the implementation as overrides. Pyright will report these issues if you enable the `reportIncompatibleMethodOverride` check.

**Comment by mirhahn** [comment]
Thanks. Is there a particular reason why that check is deactivated by default? It seems pretty essential in this case.

**Comment by erictraut** [comment]
It's pretty common in Python code to override methods in ways that are not strictly safe from a type perspective. Enabling this rule by default would generate a lot of unwelcome noise for typical pyright and pylance users.

For explicit protocol implementations specifically, I can understand why it might be more desirable to enable this check by default. To do so, we'd need to split the current diagnostic rule into two separate rules (one for explicit protocol implementations and one for all other classes) and enable the former but not the latter by default. Splitting the existing rule comes with some costs (e.g. churn for existing pyright users), so I'm not sure it's worth it. Let's leave it as is for now. We can see if we get any further feedback from users on this topic.

**Comment by mirhahn** [comment]
I see your point and for my purposes, turning the flag on is an entirely adequate solution. Thank you.

Just a quick note if you ever get back to it: there may not be need to split the rule (at least in terms of the interface) and clutter the configuration. The documentation states that the "reportIncompatibleMethodOverride" setting takes string values, so maybe setting it to "protocol" could limit the warning to situations where the method comes from an inherited protocol. This would leave open the option of extending the setting in the future to accept other such "contexts" or even lists of them if more special use cases for these warnings (e.g., checking only methods decorated with `override`) should arise.

As I said, I require none of that, so don't waste time on my account.

15.05.2023 23:22:52 Eric Traut ***@***.***>:

**Comment by erictraut** [comment]
The string value for all rules in pyright indicates the severity level, and it must be one of "none", "information", "warning", or "error". It wouldn't work to support "protocol", since that's orthogonal to severity level. I think a new rule would be required in this case.
